### DISCUSSION TITLE
Help understanding op/asset-based concurrency

### DISCUSSION THREAD

**Discussion by Zibc** [dagster-io/dagster#20220] [Q&A]
Hello,

I am trying to setup a graph to run some sequential actions.
Let's say I have 3 actions that could be either ops or assets. action 1 is leading, its output is required for action 2, same for action 3.
My current approach: I have a sensor that scans files in a directory, and for each line in this directory, yields a RunRequest, passing the line data to the op config in the run config.

The job I have is simply gluing the 3 actions together.

The catch is that, now, I want to start limiting concurrency, because some of the ops/assets make use of licenses, which only exists in limited number. I followed the "limiting concurrency" page, but failed quite miserably to use anything the run-level concurrency limits.

After some reading of the doc, I sort of figured out that this was what the doc referred to as "global concurrency limit", which does not work with SQLite ... which is the only way I can use Dagster.

So now, what I am wondering is first, how the op/asset based tag concurrency limit is supposed to work: I am not understanding how, in a job, I can trigger parallel op evaluation, and then trigger this concurrency limit: I would be very happy to read further opinions on this.

Second question is then how can I re-write how my job is ran to make use of this inner-run concurrency limit. Something like trigger multiple asset materialization of the same function; but with different configs, and coupling this in a correct way to the downstream action.

Suggestions are very much appreciated.

**Comment by alangenfeld** [comment]
> how the op/asset based tag concurrency limit is supposed to work

The database is used to coordinate the availability and claims of the limited "concurrency slots." Every Dagster run is driven by an Executor and these executors ensure that a concurrency slot is held before starting an op/asset that needs one 
https://docs.dagster.io/deployment/executors#executors

 > Second question  ...
 
 I am not able to understand what you are asking here
 
 > failed quite miserably to use anything the run-level concurrency limits
 
Sounds like you are using sqlite, are you using a daemon and [QueuedRunCoordinator](https://docs.dagster.io/_apidocs/internals#dagster._core.run_coordinator.QueuedRunCoordinator) ? That run coordinator is necesary to respecting run level concurrency settings.  

https://docs.dagster.io/deployment/run-coordinator#run-coordinators

**Reply by Zibc** [reply]
Sorry to come back to you that late, but better late than never.

Thanks for giving details about the slots management in the database.

I am using the QueuedRunCoordinator.

I also managed to get my pipelines to work after realising that one asset key correspond literally to one asset: if I re-materialise an existing asset key from another job, it will set a new value for this specific asset, not create a new standalone asset (yes, I also realise that was a flawed way of thinking)
So I then tried to make an asset factory, a bit like the op factory detailed in the doc, and that actually worked.
So I have now only one job, in one run that contains a whole list of assets and their corresponding links, using the SQLite db. That way, I could get the tag-based concurrency to work!

What I am now wondering is whether this a clean way to do things.
